qualtran/bloqs/chemistry/thc_select.py

```diff
@@ -0,0 +1,278 @@
+    from Ref. 2. This is a placeholder waiting for an actual implementation.
+    See https://github.com/quantumlib/Qualtran/issues/386.
+
+    Args:
+        num_bits_theta: Number of bits of precision for the rotation angles.
+        rotation_angles: A list of rotation angles
+
+    References:
+        [Even more efficient quantum computations of chemistry through
+            tensor hypercontraction](https://arxiv.org/pdf/2011.03494.pdf) Fig. 7.
+        [Quantum computing enhanced computational catalysis]
+        (https://arxiv.org/abs/2007.14460).
+            Burg, Low et. al. 2021. Eq. 73
+    """
+
+    num_mu: int
+    num_spin_orb: int
+    num_bits_theta: int
+    rotation_angles: Sequence[Sequence[int]]
+    optimal_kr1: int = 1
+    optimal_kr2: int = 1
+    two_body_only: bool = False
+    adjoint: bool = False
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature(
+            [
+                Register("eq_nu_mp1", bitsize=1),
+                Register("data", bitsize=self.num_bits_theta),
+                Register("sel", bitsize=self.num_mu.bit_length()),
+                Register("trg", bitsize=self.num_spin_orb // 2),
+            ]
+        )
+
+    def pretty_name(self) -> str:
+        dag = '†' if self.adjoint else ''
+        return f"In_mu-R{dag}"
+
+    def bloq_counts(self, ssa: Optional['SympySymbolAllocator'] = None) -> Set[Tuple[int, Bloq]]:
+        # from listings on page 17 of Ref. [1]
+        num_data_sets = self.num_mu + self.num_spin_orb // 2
+        if self.adjoint:
+            if self.two_body_only:
+                toff_cost_qrom = (
+                    int(np.ceil(self.num_mu / self.optimal_kr1))
+                    + int(np.ceil(self.num_spin_orb / (2 * self.optimal_kr1)))
+                    + self.optimal_kr1
+                )
+            else:
+                toff_cost_qrom = int(np.ceil(self.num_mu / self.optimal_kr2)) + self.optimal_kr2
+        else:
+            toff_cost_qrom = num_data_sets - 2
+            if self.two_body_only:
+                # we don't need the only body bit for the second application of the
+                # rotations for the nu register.
+                toff_cost_qrom -= self.num_spin_orb // 2
+        # xref https://github.com/quantumlib/Qualtran/issues/370, the cost below
+        # assume a phase gradient.
+        rot_cost = self.num_spin_orb * (self.num_bits_theta - 2)
+        return {(4 * (rot_cost + toff_cost_qrom), TGate())}
+
+
+@frozen
+class SelectTHC(Bloq):
+    r"""SELECT for THC Hamilontian.
+
+    Args:
+        num_mu: THC auxiliary index dimension $M$
+        num_spin_orb: number of spin orbitals $N$
+        angles: list of num_mu + num_spin_orb / 2 lists of rotation angles each
+            of length num_spin_orb//2. Each rotation angle should be a
+            `num_bits_theta` fixed point approximation to the angle.
+
+    Registers:
+     - mu: $\mu$ register.
```

Comment:
sorry, can you fix the formatting of these as well

Reply:
Done.